### ISSUE TITLE
family/meta/resource names

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#465]
Given the set of breaking changes to (mostly low level) interfaces in the `Meta` trait in #385. We should reconsider some of the sketchier names __before releasing a minor__:

In particular, see this comment: https://github.com/clux/kube-rs/pull/385#discussion_r565568241:

> rename `Meta` to `Resource`
> rename `Resource` to `RequestBuilder/QueryBuilder`
> rename `Family` to `RuntimeTypeInformation` (long name, but anyway users will rarely interact with it)

My personal opinions here:

1. `Meta` to `Resource` is a nice suggestion, albeit an annoying rename for every user of it (and people use it for grabbing stuff out of metadata). It would map nicely onto `k8s_openapi::Resource`, but that said, our trait kind of does both `k8s_openapi::{Resource, Metadata}` now, so it's not a huge deal to not be aligned. I am OK with keeping `Meta`, no real preference.

2. Kind of dependent on whether we go for `1`, the `Resource` to `RequestBuilder` does feel sensible, and it frees up the `Resource` name, but again a huge breaking change for people using `Resource` directly (and [people still do](https://github.com/search?l=Rust&q=kube%3A%3AResource&type=Code))

(There's also the practicality of swapping Resource from an object to a trait in one release, it might have to be a multi-part change.)

3. `Family` otoh, is __definitely not the right name__. The suggested `RuntimeTypeInformation` or `RTI` is a lot more descriptive (and less fear-inducing for people; most people will encounter it as an empty type), and there's also some precedence for the abbreviation (RTTI in C++). It's long, but it's not many places people have to specify it. But maybe there are better suggestions. The Family instance for dynamic types kind of replaces the type-erased object kind. `RuntimeExtension`? Perhaps maps a little bit onto the [dynamic types in apimachinery](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension)

**Comment by clux** [comment]
Sorry to tag you both in more bike-shed. Feel free to abstain if you are low on time.

**Comment by clux** [comment]
cc @MikailBag

**Comment by teozkr** [comment]
Haven't really read up on what `Family` does, so can't comment there (although it seems like users will need to get one to make a `Controller`?).

IMO one of the really sore spots of `kube` ATM is the confusion between `kube::Meta`, `kube::Resource`, `k8s_openapi::Metadata`, and `k8s_openapi::Resource`. Especially clearing up the `Resource` conflict is definitely worth doing..

**Comment by lfrancke** [comment]
Some comments from a user only. I don't know the implementation details very well so take it with a grain of salt.

1. I'm in favor of `Meta` to `Resource`. Yes, it's a bit annoying but it's much clearer and consistent with `k8s_openapi`
2. I have to be honest: As a user I never even saw the `kube::Resource` up until this issue so while I wasn't confused before I definitely am now :)
If I understand correctly it's the thing used by `Api` under the hood? Is there a reason for users to interact with this directly?

I'd be in favor of switching in one release but that's just me. It's changing anyway and I'd much rather have one breaking release than multiple.

And a bit of kinda related feedback about 3.:

3.I did look at the new `Meta` and I saw `Family` but I'm afraid I don't understand the comment above it at all. It might be I'm alone with this but I'm going to assume others have the same issue. `RuntimeTypeInformation` doesn't help me personally either. For me it's mostly a documentation thing at the moment before I would even be able to comment on a name.

**Comment by clux** [comment]
> Haven't really read up on what `Family` does, so can't comment there (although it seems like users will need to get one to make a `Controller`?).

The TL;DR on `Family`: it's an "optional" associated type parameter for the updated `Meta` trait to allow dynamic types to use it (without implementing `k8s_openapi` traits). There are two blanket implementations of `Meta` with two different Families:

- `impl Meta<K> for K` where `K: k8s_openapi::Metadata<..>` and `Family = ()`
- `impl Meta for DynamicObject` where `Family = GroupVersionKind` (the type information we need to fill the gap)

Most `Controller` usages won't need it. Type parameter fall away `kube_runtime` generally so should not show up much in user code. Here is hte [one exception](https://github.com/clux/kube-rs/commit/6b95f200844ef5a1a35337cff8a4fe94ca4bf880#diff-a4035f22dfb0b28537bbec49b414a056c4ee45c179e272af383f43cf16fcee39) i've seen.

**Comment by kazk** [comment]
I don't have suggestions for `Family` at the moment, but I agree we need a better name.

Renaming `Resource` is definitely worth doing.

> but again a huge breaking change for people using `Resource` directly (and [people still do](https://github.com/search?l=Rust&q=kube%3A%3AResource&type=Code))

The search result shows 8.
- 2 are from <https://github.com/slinkydeveloper/extending-kubernetes-api-in-process-poc> that uses a hacked `kube` according to their README.
- 2 are comments from vendored `kube-rs`
- 1 is invalid match for `kube_resource` parameter from <https://github.com/flodiebold/cautious-tribble> that doesn't use `kube`.
- 1 is <https://github.com/nikitavbv/my-k8s-dashboard> which is pretty new and WIP. They're [still learning `kube`](https://github.com/nikitavbv/my-k8s-dashboard/blob/7af71a44d742b4a10274b2a131c98c32c7258445/src/usage.rs#L46-L47). `kube::Resource::all::<Pod>()`
- 1 is <https://github.com/stephenc/k8s-gcr-auth-helper> and has [`kube::Resource::all::<k8s_openapi::api::core::v1::Event>()`](https://github.com/stephenc/k8s-gcr-auth-helper/blob/86a86103d9238072d6a33116d9c39e3cf57d52a6/src/watch.rs#L51)
- 1 is a comment in <https://github.com/babylonhealth/shipcat/blob/ded1a2fe6f30fb9640b068639514f63d848d1869/shipcat_cli/src/kubeapi.rs#L112> by @clux. But this does import `kube::Resource` and uses it.

(It's difficult to find usage with GitHub search because this misses imports like `kube::{Api, Resource}` and [`kube Resource`](https://github.com/search?q=kube+Resource&type=Code&l=rust) is too noisy.)

**Comment by clux** [comment]
Yeah, it was a quick search, and while there are probably more if we can search for wildcard imports, the scope of the breakage might not be that big.

At any rate, it seems like all of you so far are pro renaming `Resource`, so maybe we should just do it. There's no time like semver < 1.

**Comment by clux** [comment]
Going to start with the obvious thing to rename; `Family` in #472.

**Comment by clux** [comment]
Onto part two: `Resource` -> `RequestBuilder`. PR coming up.

**Comment by clux** [comment]
Not sure about `RequestBuilder` as a new name for `Resource` as it currently stands. It currently doesn't just build requests, has a bunch of state between calls, has a separate builder `DynamicResource` to fill in this state (effectively making it a `ResourceBuilderBuilder` if we were to rename).

It's possible we can instead take advantage `Meta` more, and instead rename `Resource` to `Request`, and pull out all its state. Then maybe we could get something like `Request::create<Type>(params);`, where the `Type` implements `Meta` and pull the data either out of its `GroupVersionKind` or the `k8s_openapi::Resource` associated constants at `Request` call time.

Might be a little iffy with namespace vars (we probably have to make it required for a Namespace scoped resource, and make separate getters/watchers for `--all`), but might make the struct a little more focused, which is ultimately why the naming never felt right; it's hard to define what it does singularly.

Anyway, will experiment a bit in a separate PR.

**Comment by teozkr** [comment]
Now that we have `tower` (decoupling us from `reqwest`/`hyper`) and dynamic-aware `Api`, is it worth keeping `Resource` around (as part of the public API) at all?

**Comment by clux** [comment]
> Now that we have `tower` (decoupling us from `reqwest`/`hyper`) and dynamic-aware `Api`, is it worth keeping `Resource` around (as part of the public API) at all?

Very possible. But something needs to at the very least do the `http::Request` making logic, focusing on making that nice first.

I have renamed `Resource` to `Request` in #474 along with some other changes to make this possibly more palatable, and have less state carrying. I think this PR has a little more potential than the previous one (though it needs some work still).

**Comment by clux** [comment]
Final PR continuing after #474 in #478. Planning merging both tomorrow.

The big improvements;  the new `Request` (old Resource) now being basically not stateful + the lifting of the `url_path` method into the new `Resource` trait (old `Meta`) are kind of the key changes, the rest is just fallout from those restructurings and renames.

I don't think it looks bad, but it will almost certainly be a bit annoying to do the upgrade when `kube::api::Resource` changes meaning in the next version, so will have to write a bigger disclaimer in the CHANGELOG.
